File: prg/2048/game.py
"""
Hlavní třída hry - řídí průběh a logiku herního kola.
"""
from board import GameBoard


class Game:
    """Třída řídící hru 2048 pro dva hráče."""
    
    def __init__(self):
        """Inicializuje novou hru."""
        self.board = GameBoard()
        self.current_player = 'dog'  # Začíná hráč s psy
        self.dog_won = False
        self.cat_won = False
        self.game_over = False
        self.moves = 0
    
    def switch_player(self):
        """Přepne na dalšího hráče."""
        self.current_player = 'cat' if self.current_player == 'dog' else 'dog'
    
    def get_current_player_name(self) -> str:
        """Vrací jméno aktuálního hráče."""
        return "Psi" if self.current_player == 'dog' else "Kočky"
    
    def process_move(self, direction: str) -> bool:
        """
        Zpracuje pohyb hráče.
        
        Args:
            direction: Směr pohybu ('up', 'down', 'left', 'right')
        
        Returns:
            True pokud byl tah platný, False jinak.
        """
        # Kontrola platnosti směru
        if direction not in ['up', 'down', 'left', 'right']:
            return False
        
        # Provede pohyb
        if not self.board.move(direction, self.current_player):
            return False  # Tah nebyl možný
        
        self.moves += 1
        
        # Kontrola vítězství
        if self.board.has_player_won(self.current_player):
            self.game_over = True
            if self.current_player == 'dog':
                self.dog_won = True
            else:
                self.cat_won = True
            return True
        
        # Kontrola konce hry
        if self.board.is_game_over():
            self.game_over = True
            return True
        
        # Přepne na dalšího hráče
        self.switch_player()
        return True
    
    def display_state(self):
        """Zobrazí aktuální stav hry."""
        self.board.display()
        
        dog_level = self.board.get_highest_level('dog')
        cat_level = self.board.get_highest_level('cat')
        
        print(f"Psi (nejvyšší úroveň): {dog_level}/3")
        print(f"Kočky (nejvyšší úroveň): {cat_level}/3")
        print(f"Celkový počet tahů: {self.moves}")
        print(f"Na tahu: {self.get_current_player_name()}\n")
    
    def display_rules(self):
        """Zobrazí pravidla hry."""
        print("\n" + "=" * 50)
        print("PRAVIDLA HRY 2048")
        print("=" * 50)
        print("""
Dva hráči si střídají tahy:
1. Hráč s psy (štěně → pes → vlk)
2. Hráč s kočkami (kotě → kočka → lev)

SLUČOVÁNÍ ZVÍŘAT:
- Stejný typ + stejná úroveň = evoluce (vyšší úroveň)
- Různé typy + stejná úroveň = nic (pokud je to typ aktuálního hráče)
- Různá úroveň = vyšší úroveň vítězí

CÍL:
- Vytvořit vlka (pro psy) nebo lva (pro kočky)
- Nebo mít nejvyšší úroveň když už žádné tahy nejsou

OVLÁDÁNÍ:
- up, down, left, right = pohyb
- undo = vrátit poslední tah
- quit = konec hry

""")
        print("=" * 50 + "\n")
    
    def get_winner_name(self) -> str:
        """Vrací jméno vítěze."""
        if self.dog_won:
            return "Psi"
        elif self.cat_won:
            return "Kočky"
        else:
            dog_level = self.board.get_highest_level('dog')
            cat_level = self.board.get_highest_level('cat')
            if dog_level > cat_level:
                return "Psi"
            elif cat_level > dog_level:
                return "Kočky"
            else:
                return "Remíza"
    
    def display_game_over(self):
        """Zobrazí konečný stav hry."""
        print("\n" + "=" * 50)
        print("KONEC HRY")
        print("=" * 50)
        
        if self.dog_won:
            print("🐺 PSI ZVÍTĚZILI! 🐺")
        elif self.cat_won:
            print("🦁 KOČKY ZVÍTĚZILY! 🦁")
        else:
            dog_level = self.board.get_highest_level('dog')
            cat_level = self.board.get_highest_level('cat')
            
            if dog_level > cat_level:
                print("🐕 PSI ZVÍTĚZILI (vyšší úroveň)! 🐕")
            elif cat_level > dog_level:
                print("😺 KOČKY ZVÍTĚZILY (vyšší úroveň)! 😺")
            else:
                print("⚖️ REMÍZA! ⚖️")
        
        print(f"\nKonečné úrovně:")
        print(f"  Psi:    {self.board.get_highest_level('dog')}/3")
        print(f"  Kočky:  {self.board.get_highest_level('cat')}/3")
        print(f"  Tahů:   {self.moves}")
        print("=" * 50 + "\n")

File: prg/2048-go/game/game.go
// Package game implementuje hlavní logiku hry
package game

import (
	"fmt"

	"2048-go/animal"
	"2048-go/board"
)

// Game řídí průběh hry
type Game struct {
	Board         *board.Board
	CurrentPlayer animal.AnimalType
	DogWon        bool
	CatWon        bool
	GameOver      bool
	Moves         int
}

// NewGame vytvoří novou hru
func NewGame() *Game {
	return &Game{
		Board:         board.NewBoard(),
		CurrentPlayer: animal.Dog, // Začíná hráč s psy
		DogWon:        false,
		CatWon:        false,
		GameOver:      false,
		Moves:         0,
	}
}

// SwitchPlayer přepne na dalšího hráče
func (g *Game) SwitchPlayer() {
	if g.CurrentPlayer == animal.Dog {
		g.CurrentPlayer = animal.Cat
	} else {
		g.CurrentPlayer = animal.Dog
	}
}

// GetCurrentPlayerName vrací jméno aktuálního hráče
func (g *Game) GetCurrentPlayerName() string {
	if g.CurrentPlayer == animal.Dog {
		return "Psi"
	}
	return "Kočky"
}

// ProcessMove zpracuje pohyb hráče
func (g *Game) ProcessMove(direction string) bool {
	// Kontrola platnosti směru
	if direction != "up" && direction != "down" && direction != "left" && direction != "right" {
		return false
	}

	// Provede pohyb
	if !g.Board.Move(direction, g.CurrentPlayer) {
		return false // Tah nebyl možný
	}

	g.Moves++

	// Kontrola vítězství
	if g.Board.HasPlayerWon(g.CurrentPlayer) {
		g.GameOver = true
		if g.CurrentPlayer == animal.Dog {
			g.DogWon = true
		} else {
			g.CatWon = true
		}
		return true
	}

	// Kontrola konce hry
	if g.Board.IsGameOver() {
		g.GameOver = true
		return true
	}

	// Přepne na dalšího hráče
	g.SwitchPlayer()
	return true
}

// DisplayState zobrazí aktuální stav hry
func (g *Game) DisplayState() {
	g.Board.Display()

	dogLevel := g.Board.GetHighestLevel(animal.Dog)
	catLevel := g.Board.GetHighestLevel(animal.Cat)

	fmt.Printf("Psi (nejvyšší úroveň): %d/3\n", dogLevel)
	fmt.Printf("Kočky (nejvyšší úroveň): %d/3\n", catLevel)
	fmt.Printf("Celkový počet tahů: %d\n", g.Moves)
	fmt.Printf("Na tahu: %s\n\n", g.GetCurrentPlayerName())
}

// DisplayRules zobrazí pravidla hry
func (g *Game) DisplayRules() {
	fmt.Println("==================================================")
}

// GetWinnerName vrací jméno vítěze
func (g *Game) GetWinnerName() string {
	if g.DogWon {
		return "Psi"
	} else if g.CatWon {
		return "Kočky"
	} else {
		dogLevel := g.Board.GetHighestLevel(animal.Dog)
		catLevel := g.Board.GetHighestLevel(animal.Cat)
		if dogLevel > catLevel {
			return "Psi"
		} else if catLevel > dogLevel {
			return "Kočky"
		} else {
			return "Remíza"
		}
	}
}

// DisplayGameOver zobrazí konečný stav hry
func (g *Game) DisplayGameOver() {
	fmt.Println("\n" + "==================================================")
	fmt.Println("KONEC HRY")
	fmt.Println("==================================================")

	if g.DogWon {
		fmt.Println("PSI")
	} else if g.CatWon {
		fmt.Println("KOČKY")
	} else {
		dogLevel := g.Board.GetHighestLevel(animal.Dog)
		catLevel := g.Board.GetHighestLevel(animal.Cat)

		if dogLevel > catLevel {
			fmt.Println("PSI")
		} else if catLevel > dogLevel {
			fmt.Println("KOČKY")
		} else {
			fmt.Println("REMÍZA")
		}
	}

	fmt.Println("\nKonečné úrovně:")
	fmt.Printf("  Psi:    %d/3\n", g.Board.GetHighestLevel(animal.Dog))
	fmt.Printf("  Kočky:  %d/3\n", g.Board.GetHighestLevel(animal.Cat))
	fmt.Printf("  Tahů:   %d\n", g.Moves)
	fmt.Println("==================================================")
}
